### DISCUSSION TITLE
Why not C++11 as the minimum requirement?

### DISCUSSION THREAD

**Discussion by stefanos82** [ocornut/imgui#3890] [Old/Archived discussions]
Out of sheer curiosity, how come it's not C++11 the minimum requirement to coding implementation?

There are things that I personally think could help with the project's development as a whole.

I'm not saying to use the entire C++11 standard; far from it!

Only those features the project may actually need.

For instance, I read a bit of the code and wondered how much cleaner the logic would be with lambdas wherever applicable.

The same goes for with the use of `constexpr`, `auto`, and maybe somewhere the `decltype`.

Of course, you know best why you choose to stick with a subset of C++; let's call it C++Lite...-ish :laughing: 

Can you enlighten me please with your reasoning so I learn more how an experienced Graphics Developer thinks?

**Comment by eliasdaler** [comment]
I love modern C++ a lot, but I think that Dear ImGui won't benefit from it much.
Standard containers or `std::string` aren't used in code, so new additions in standard library won't help it much.
Dear ImGui doesn't do dynamic allocations (maybe in some rare places), so `std::unique_ptr` isn't needed.

The only thing I can think of is using `enum class` instead of `enum` so that enum values don't leak into global namespace and there's no implicit casts between different enum types.

If there are any examples where modern C++ will make some code better, I'd like to see them.

But otherwise, I think that it would be good if someone could write a "wrapper" library around ImGui which implements modern C++ API with template magic, RAII wrappers for `Push`/`Pop` stuff, lambdas for callbacks, etc. 
But implementing this in Dear ImGui's repo would increase the code size substantially, which is not very good.

**Reply by stefanos82** [reply]
I thought I was clear enough with 

> Only those features the project may actually need.

Looks like I wasn't!

In other words, if there is a C++11 feature that I find it useful and frees me from implementing it in the form of a macro, why not use that existing standard feature instead?

I don't believe by changing a compiler flag to use `-std=c++11` in place of `-std=c++03` would add any extra code apart from what you are using, but then again I could be wrong.

**Comment by ocornut** [comment]
Stephanos thanks for asking and being curious.

Traditionally the reason for that friction in pro game development is that the need to haves game running on non-traditional architectures means compilers are rarely standard or latest versions. Many "modern-ish" C++ features also have performances properties (either runtime or compilation-time) that not-acceptable for AAA developers, or dependencies that are not acceptable either. 

Lots of established tech wouldn't be able to use Dear ImGui if it required recent versions or dragged dependencies. Many AAA gamedev would never even consider dear imgui if it used e.g. std::string (of course that's not the core of your question, but it stills gives an element of answer). Lots of homebrew projects are also stuck with old toolsets and don't want to upgrade. New programmers are notoriously less agile, so less able to change toolsets.

As we are shipping a LIBRARY, is not our role to require people to upgrade especially as in many situations they simply cannot. When I'm working on a private project which I have control of, I can use newer features, as many game developers today I uses features from C++11 and onwards when I can, but Dear ImGui is less a place for it.

That said I think we will probably up the compiler requirements of Dear ImGui sometimes this year to align with a subset of C++11, so we can benefit from a few extra compilers and low-level header features.

> Only those features the project may actually need. 

Well, what are they?

**Reply by stefanos82** [reply]
> Stephanos thanks for asking and being curious.
> 
> Traditionally the reason for that friction in pro game development is that the need to haves game running on non-traditional architectures means compilers are rarely standard or latest versions. Many "modern-ish" C++ features also have performances properties (either runtime or compilation-time) that not-acceptable for AAA developers, or dependencies that are not acceptable either.
> 

I understand.

> Lots of established tech wouldn't be able to use Dear ImGui if it required recent versions or dragged dependencies. Many AAA gamedev would never even consider dear imgui if it used e.g. std::string (of course that's not the core of your question, but it stills gives an element of answer). Lots of homebrew projects are also stuck with old toolsets and don't want to upgrade. New programmers are notoriously less agile, so less able to change toolsets.
>

I never mentioned anything about STL elements, such as `std::string`.

> As we are shipping a LIBRARY, is not our role to require people to upgrade especially as in many situations they simply cannot. When I'm working on a private project which I have control of, I can use newer features, as many game developers today I uses features from C++11 and onwards when I can, but Dear ImGui is less a place for it.
> 
> That said I think we will probably up the compiler requirements of Dear ImGui sometimes this year to align with a subset of C++11, so we can benefit from a few extra compilers and low-level header features.
>

That's exactly what I have asked in the original question but for some reason it ended up being about `std::string` as if I mentioned the use of C++ Standard Libraries...I'm very confused right now.

I guess when people hear or read the word `standard` accompanied by C++ automatically gets interpreted as Standard Libraries?

I'm referring at ANSI / ISO C++ standard.

> > Only those features the project may actually need.
> 
> Well, what are they?

`constexpr`, `auto`, `decltype`, lambda functions, `static_assert`, list initialization to prevent narrowing (e.g.: int val {3.0} // error: no double to int value conversion)

**Comment by ocornut** [comment]
You are right, but also unfortunately modern C++ insistence in over-complicatedly making things "libraries" (part of headers) instead of defining them in the language, means that some features which should be language features instead depends on C++ headers (I think that's the case for initializer list), and those are becoming increasingly hard to trust long-term.

Right now my "minimum build target" in mind has been VS2010 and I worked with those constraints. May up that to 2013 soon since that's the time were they added better support for C++11 features.

My question was most "do you have specific use in mind" for use?
dear imgui does not use callback very much and doesn't seem to benefit from lambda.
From my pov the most useful things would be enum, range-for.

**Reply by stefanos82** [reply]
> My question was most "do you have specific use in mind" for use?
> dear imgui does not use callback very much and doesn't seem to benefit from lambda.
> From my pov the most useful things would be enum, range-for.

I guess you can continue with what you are currently using and that should do; no need to add an extra layer of unnecessary complexity.

Maybe a user / company survey about the compiler they choose to use could be helpful to decide a future C++11 support?

**Comment by iocafe** [comment]
I often work with development environments tested and created for specific hardware platform more than 10 years ago, sometimes much more. Industrial stuff lives often long time. Rewriting and testing board support, etc to production level, is expensive and lengthy project. In practice requiring upgrade to newest tools would prevent using the library, or freeze it to latest backwards compatible version. This would not only prevent update library in old designs, but also in new ones which need to share code with the old ones. In my view, benefits of using C++11 are cosmetic. Small part of code would look neater, but that is all.
